Fixes #7175 - Allow to search for email address in user management

Abijeet opened the pull request:
Greetings,

@MorrisJobke  - Please find the merge request that fixes #7175 

I've done the following,

1. Updated the query to fetch the users in users > everyone tab.
2. Updated the query to fetch the users in users > admin tab.
3. Tested to ensure that the disabled users are also being fetched.

Tested with SQLite.

Signed-off-by: Abijeet <abijeetpatro@gmail.com>

Abijeet commented:
Not sure why the test case failed, please let me know if there is anything I've to do or change.

MorrisJobke commented:
> Not sure why the test case failed, please let me know if there is anything I've to do or change.

Fixed in master: #7433

blizzz commented:
It works, however I'D appreciate "unit" tests in `tests/lib/User/DatabaseTest.php` to ensure this keeps working in the future. Basically `testSearch()` from the inherited "Backend" class can be extended to look for the email.

Abijeet commented:
@blizzz - Thank you. I'll add the test cases some time today.

Abijeet committed:
Adds search by email function on the users screen.

Fixes #7175.

- Updated the query to fetch the users in users > everyone tab.
- Updated the query to fetch the users in users > admin tab.
- Tested to ensure that the disabled users are also being fetched.
- Added test cases.

Signed-off-by: Abijeet <abijeetpatro@gmail.com>

Abijeet commented:
@blizzz , @MorrisJobke  - 

Couple of things,

1. I'm not sure if I've implemented the test cases properly. Please let me know if I've to change anything here.
2. There was an extra space in the `.htaccess` that Netbeans trimmed off, let me know if I should undo that.

Abijeet commented:
Also I noticed one more thing, searching with `_` in LIKE isn't working. 

So for example, if there is a user with `uid` as **admin_user**. 

If you search for this user in the view with the following search term - **admin_**, you'll not get any value in the response. This is because `escapeLikeParameter` function adds a `\` before the `_` (Ref: https://github.com/nextcloud/server/blob/master/lib/private/DB/Connection.php#L410). Tested on SQLite. 

Not sure if this is a known issue.

MorrisJobke commented on .htaccess:
```diff
@@ -23,7 +23,7 @@
     Header set Cache-Control "max-age=15778463"
   </FilesMatch>
-  
+
```
This is fine.

MorrisJobke commented on .htaccess:
```diff
@@ -77,4 +77,4 @@ AddDefaultCharset utf-8
 <IfModule pagespeed_module>
   ModPagespeed Off
-</IfModule>
+</IfModule>
```
That change should be undone. We usually have a new line at the end of files.

Abijeet replied:
Fixed, added the newline back. Will keep in mind for future.

Abijeet committed:
Added newline to end of htaccess file

Signed-off-by: Abijeet <abijeetpatro@gmail.com>

blizzz commented:
> This is because escapeLikeParameter function adds a \ before the _ (Ref: https://github.com/nextcloud/server/blob/master/lib/private/DB/Connection.php#L410). Tested on SQLite.

This was added for security reasons, since % and _ act as wildcards in SQL. The \ is necessary to escape them so user input does not act  like a wildcard. For sqlite to work it requires an additional `ESCAPE '\'` however. This is not the case?

icewind1991 committed:
refactor user searching
add additional user searching tests

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
Took the opportunity to refactor the user searching code and fixed the `_` issue

@blizzz please re-review

blizzz commented on lib/private/User/Database.php:
```diff
@@ -189,23 +196,29 @@ public function getDisplayName($uid) {
+			->from('users', 'u')
+			->leftJoin('u', 'preferences', 'p', $query->expr()->andX(
+				$query->expr()->eq('userid', 'uid')),
+				$query->expr()->eq('appid', new Literal('settings')),
```
Literal cannot be hidden from consumers?

icewind1991 replied:
All plain strings are used as column names

blizzz replied:
I mean whether everywhere where we use strings in queries they need to be instances of ``Literal`` or whether it can be abstracted away?

icewind1991 replied:
We need to use `Literal` to explicitly tell the query builder that the string is safe to use

blizzz replied:
ok

blizzz commented:
@icewind1991 just the other question, looks good otherwise and is working

icewind1991 committed:
reuse query builder

Signed-off-by: Robin Appelman <robin@icewind.nl>

blizzz commented:
needs another reviewer

MorrisJobke reviewed (approved):
Tested and works 👍

MorrisJobke commented on lib/public/DB/QueryBuilder/IFunctionBuilder.php:
```diff
@@ -71,4 +71,13 @@ public function substring($input, $start, $length = null);
+	 *
+	 * @param mixed $field
+	 * @return IQueryFunction
+	 * @since 13.0.0
```
See #8696

Abijeet commented:
Finally :)
